src/game/emutest.c

```diff
@@ -0,0 +1,168 @@
+#include <PR/R4300.h>
+#include <ultra64.h>
+#include <string.h>
+#include "types.h"
+
+extern OSMesgQueue gSIEventMesgQueue;
+extern u8 __osContPifRam[];
+extern u8 __osContLastCmd;
+extern void __osSiGetAccess(void);
+extern void __osSiRelAccess(void);
+
+enum Emulator gEmulator = EMU_CONSOLE;
+
+__attribute__((aligned(8)))
+static const u32 check_count_factor_asm[] = {
+    0x40084800u, // MFC0 T0, COUNT
+    0x40094800u, // MFC0 T1, COUNT
+    0x03E00008u, // JR RA
+    0x01281023u  // SUBU V0, T1, T0
+};
+
+static inline u32 check_count_factor() {
+    const u32 saved = __osDisableInt();
+    const u32 cf = ((u32 (*)(void))check_count_factor_asm)();
+    __osRestoreInt(saved);
+    return cf;
+}
+
+static inline enum Emulator get_pj64_version() {
+    // When calling this function, we know that the emulator is some version of Project 64,
+    // and it isn't using the PJ64 4.0 interpreter core. Figure out which version it is.
+    
+    // PJ64 4.0 dynarec core doesn't update the COUNT register correctly within recompiled functions
+    if (check_count_factor() == 0) {
+        return EMU_PROJECT64_4;
+    }
+
+    // Instead of implementing this PIF command correctly, PJ64 versions prior to 3.0 just have
+    // a set of hardcoded values for some requests. At least one of these hardcoded values has a
+    // typo in it, making it give an incorrect result.
+    __osSiGetAccess();
+    u32 *pifRam32 = (u32*)__osContPifRam;
+    for (s32 i = 0; i < 15; i++) pifRam32[i] = 0;
+    pifRam32[15] = 2;
+
+    const u8 cicTest[] = {
+        0x0F, 0x0F,
+        0xEC, 0x3C, 0xB6, 0x76, 0xB8, 0x1D, 0xBB, 0x8F,
+        0x6B, 0x3A, 0x80, 0xEC, 0xED, 0xEA, 0x5B
+    };
+
+    memcpy(&__osContPifRam[46], cicTest, 17);
+
+    __osSiRawStartDma(OS_WRITE, __osContPifRam);
+    osRecvMesg(&gSIEventMesgQueue, NULL, OS_MESG_BLOCK);
+    __osContLastCmd = 254;
+
+    __osSiRawStartDma(OS_READ, __osContPifRam);
+    osRecvMesg(&gSIEventMesgQueue, NULL, OS_MESG_BLOCK);
+    const u8 pifCheck = __osContPifRam[54];
+    __osSiRelAccess();
+
+    // check for the typo'd byte
+    return (pifCheck == 0xB0) ? EMU_PROJECT64_1_OR_2 : EMU_PROJECT64_3;
+}
+
+static u8 check_cache_emulation() {
+    // Disable interrupts to ensure that nothing evicts the variable from cache while we're using it.
+    u32 saved = __osDisableInt();
+    // Create a variable with an initial value of 1. This value will remain cached.
+    volatile u8 sCachedValue = 1;
+    // Overwrite the variable directly in RDRAM without going through cache.
+    // This should preserve its value of 1 in dcache if dcache is emulated correctly.
+    *(u8*)(K0_TO_K1(&sCachedValue)) = 0;
+    // Read the variable back from dcache, if it's still 1 then cache is emulated correctly.
+    // If it's zero, then dcache is not emulated correctly.
+    const u8 cacheEmulated = sCachedValue;
+    // Restore interrupts
+    __osRestoreInt(saved);
+    return cacheEmulated;
+}
+
+f32 round_double_to_float(f64 v) {
```

Comment:
Having this functon inside emutest.c might make gcc optimize this cast. It will break VC detection.

Reply:
What if I just add `__attribute__((externally_visible))`? Is that sufficient?

Reply:
`ALWAYS_INLINE` could work but it throws a warning to me.

Reply:
Pretty sure that's the opposite of what we want. If it can inline, then it can optimize out the cast. We want it to _not_ inline it. `__attribute__((externally_visible))` is one way to ensure that can't happen.

Reply:
imo the only "safe" way to accomplish that is guaranteed to work is to move this function to a different compilation unit as it was done previously.

Reply:
Ah, according to the gcc docs, adding `asm ("");` also works because it makes the compiler assume the function has side effects, and thus will never optimize out a call to it.

I can just move it to another translation unit though (however, this will break Wii VC if compiled with `-fwhole-program`)

Reply:
Actually, `__attribute__((externally_visible))` will prevent that. So I'll do both.